include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -891,6 +891,107 @@ namespace internal
         }
     }
 
+    /**
+     * Same as above but with hessians.
+     */
+#ifndef DEBUG
+    DEAL_II_ALWAYS_INLINE
+#endif
+    static void
+    do_backward_hessians(
+      const unsigned int                    n_components,
+      const dealii::AlignedVector<Number2> &transformation_matrix,
+      const bool                            add_into_result,
+      Number *                              values_in,
+      Number *                              values_out,
+      const unsigned int                    basis_size_1_variable =
+        dealii::numbers::invalid_unsigned_int,
+      const unsigned int basis_size_2_variable =
+        dealii::numbers::invalid_unsigned_int)
```

Comment:
I must admit that I do not like to duplicate the `do_backward` function for the sake of bypassing the assumption of a zero somewhere in the `values` call in that function. We should make the interfaces uniform by just providing an `apply` function in the tensor product evaluators with an `enum` informing us about special cases, see also #7039.

Reply:
I agree that we should try not to introduce an additional specialization for hessian. But let's not remove the `values`/`gradients`/`hessians`-function calls now from all places (I can put it on my TODO list). I have introduced as suggested an enum (name is up to discussion) and use this now in the `FEEvaluationImplBasisChange` class. What do you think, @kronbichler?

Reply:
Absolutely, there is no need to clean things up now, neither need we remove the values/gradients/hessian calls now. But I was thinking of providing all evaluators with a `static void apply` function and uniform template parameters along the variants. Or, even better, there is a single code to run through the tensor products and just different 1D matrix-vector products.